saw-central/src/SAWCentral/Proof.hs

```diff
@@ -2030,40 +2029,37 @@ propApply ::
   Prop {- ^ propsition to apply (usually a quantified and/or implication term) -} ->
   Prop {- ^ goal to apply the proposition to -} ->
   IO (Maybe [Either Term Prop])
-propApply sc rule goal = applyFirst (asPiLists (unProp rule))
+propApply sc rule goal = applyFirst =<< asPiLists (unProp rule)
   where
 
     applyFirst [] = pure Nothing
     applyFirst ((ruleArgs, ruleConcl) : rest) =
-      do result <- scMatch sc ruleConcl (unProp goal)
+      do result <- scMatch sc ruleArgs ruleConcl (unProp goal)
          case result of
            Nothing -> applyFirst rest
            Just inst ->
-             do let inst' = [ Map.lookup i inst | i <- take (length ruleArgs) [0..] ]
-                dummy <- scUnitType sc
-                let mkNewGoals (Nothing : mts) ((nm, prop) : args) =
-                      do c0 <- instantiateVarList sc 0 (map (fromMaybe dummy) mts) prop
-                         mp <- termToMaybeProp sc c0
-                         case mp of
-                           Nothing ->
-                             fail ("goal_apply: could not find instantiation for " ++ show nm)
-                           Just p ->
-                             do cs <- mkNewGoals mts args
-                                return (Right p : cs)
-                    mkNewGoals (Just tm : mts) (_ : args) =
-                      do cs <- mkNewGoals mts args
-                         return (Left tm : cs)
-                    mkNewGoals _ _ = return []
-
-                newgoalterms <- mkNewGoals inst' (reverse ruleArgs)
-                return (Just (reverse newgoalterms))
-
-    asPiLists :: Term -> [([(Text, Term)], Term)]
+             do let mkNewGoal ec =
```

Comment:
It might be worth giving this a type signature to make it more obvious what the types of the `Left` and `Right` cases are.

Reply:
Good idea.

Reply:
See 8ae0f7394780de0d124f110be8b979880ed048ba. I added a type annotation for `applyFirst` as well.